sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobBatch.java

```diff
@@ -0,0 +1,413 @@
+
+    private final ClientLogger logger = new ClientLogger(BlobBatch.class);
+
+    private final String accountUrl;
+    private final HttpPipeline batchPipeline;
+
+    private final Deque<Mono<? extends Response>> batchOperationQueue;
+    private final List<ByteBuffer> batchRequest;
+    private final Map<Integer, BlobBatchOperationResponse> batchMapping;
+
+    private final AtomicInteger contentId;
+    private final String batchBoundary;
+
+    private BlobBatchType batchType;
+
+    public BlobBatch(BlobServiceClient client) {
+        this(client.getAccountUrl(), client.getHttpPipeline());
+    }
+
+    public BlobBatch(BlobServiceAsyncClient client) {
+        this(client.getAccountUrl(), client.getHttpPipeline());
+    }
+
+    BlobBatch(String accountUrl, HttpPipeline pipeline) {
+        this.contentId = new AtomicInteger(0);
+        this.batchBoundary = String.format(BATCH_BOUNDARY_TEMPLATE, UUID.randomUUID());
+
+        this.accountUrl = accountUrl;
+
+        boolean batchHeadersPolicySet = false;
+        List<HttpPipelinePolicy> policies = new ArrayList<>();
+        for (int i = 0; i < pipeline.getPolicyCount(); i++) {
+            HttpPipelinePolicy policy = pipeline.getPolicy(i);
+
+            // Don't want these policies in the dummy pipeline.
+            if (policy instanceof HttpLoggingPolicy
+                || policy instanceof RetryPolicy
+                || policy instanceof RequestRetryPolicy
+                || policy instanceof UserAgentPolicy) {
+                continue;
+            }
+
+            if (policy instanceof SharedKeyCredentialPolicy) {
+                batchHeadersPolicySet = true;
+                // The batch policy needs to be added before the SharedKey policy to run preparation cleanup.
+                policies.add(new BatchHeadersPolicy());
+            }
+
+            policies.add(pipeline.getPolicy(i));
+        }
+
+        if (!batchHeadersPolicySet) {
+            policies.add(new BatchHeadersPolicy());
+        }
+
+        this.batchPipeline = new HttpPipelineBuilder()
+            .policies(policies.toArray(new HttpPipelinePolicy[0]))
+            .httpClient(new BatchClient(this::sendCallback))
+            .build();
+
+        this.batchOperationQueue = new ConcurrentLinkedDeque<>();
+        this.batchRequest = new ArrayList<>();
+        this.batchMapping = new ConcurrentHashMap<>();
+    }
+
+    /**
+     * Adds a delete blob operation to the batch.
+     *
+     * @param containerName The container of the blob.
+     * @param blobName The name of the blob.
+     * @return a {@link Response} that will be used to associate this operation to the response when the batch is
+     * submitted.
+     * @throws UnsupportedOperationException If this batch has already added an operation of another type.
+     */
+    public Response<Void> delete(String containerName, String blobName) {
+        return deleteHelper(buildClient(containerName, blobName), null, null);
+    }
```

Comment:
If this adds an operation (but doesn't actually perform it), it seems weird that this returns `Response<Void>`?

Reply:
The planned implementation is a little round about where the submission will return a response that becomes populated once the batch submission returns. Could change the response type to a `Future` to better indicate that this doesn't do anything yet and document that the batch must be submitted via `submitBatch` before the response will have a value.

Reply:
I think Mono instead of Future would be preferable so we don't introduce a separate paradigm just for this. It might be ok to have this as an "async only" feature where sync customers who want to use it just have to block because I agree that there should be some sort of type-enforced means of communicating that the return value isn't valid until the batch completes.